core/src/test/java/kafka/testkit/KafkaClusterTestKit.java

```diff
@@ -214,61 +249,30 @@ metaProperties, config, new MetadataRecordSerde(), metadataPartition, KafkaRaftS
-                    props.put(KafkaConfig$.MODULE$.ListenersProp(), listeners(node.id()));
-                    props.put(KafkaConfig$.MODULE$.InterBrokerListenerNameProp(),
-                        nodes.interBrokerListenerName().value());
-                    props.put(KafkaConfig$.MODULE$.ControllerListenerNamesProp(),
-                        "CONTROLLER");
-
-                    setupNodeDirectories(baseDirectory, node.metadataDirectory(),
-                        node.logDataDirectories());
-
-                    // Just like above, we set a placeholder voter list here until we
-                    // find out what ports the controllers picked.
-                    props.put(RaftConfig.QUORUM_VOTERS_CONFIG, uninitializedQuorumVotersString);
-                    props.putAll(node.propertyOverrides());
-                    KafkaConfig config = new KafkaConfig(props, false, Option.empty());
-
-                    String threadNamePrefix = String.format("broker%d_", node.id());
-                    MetaProperties metaProperties = MetaProperties.apply(nodes.clusterId().toString(), node.id());
-                    TopicPartition metadataPartition = new TopicPartition(KafkaRaftServer.MetadataTopic(), 0);
-                    KafkaRaftManager<ApiMessageAndVersion> raftManager;
-                    if (raftManagers.containsKey(node.id())) {
-                        raftManager = raftManagers.get(node.id());
-                    } else {
-                        raftManager = new KafkaRaftManager<>(
-                            metaProperties, config, new MetadataRecordSerde(), metadataPartition, KafkaRaftServer.MetadataTopicId(),
-                            Time.SYSTEM, new Metrics(), Option.apply(threadNamePrefix), connectFutureManager.future);
-                        raftManagers.put(node.id(), raftManager);
+                    JointServer jointServer = jointServers.get(node.id());
+                    if (jointServer == null) {
+                        jointServer = new JointServer(createNodeConfig(node),
+                                MetaProperties.apply(nodes.clusterId().toString(), node.id()),
+                                Time.SYSTEM,
+                                new Metrics(),
+                                Option.apply(String.format("broker%d_", node.id())),
+                                connectFutureManager.future,
+                                faultHandlerFactory);
+                        jointServers.put(node.id(), jointServer);
+                    }
+                    BrokerServer broker = null;
+                    try {
+                        broker = new BrokerServer(
+                                jointServer,
+                                JavaConverters.asScalaBuffer(Collections.<String>emptyList()).toSeq());
+                    } catch (Throwable e) {
+                        log.error("Error creating broker {}", node.id(), e);
+                        if (broker != null) broker.shutdown();
```

Comment:
nit: similar comment as before. I think JointServer still has resources that need to be cleaned up even if it doesn't get started (e.g. `Metrics`). Same thing for the controller above.

Reply:
I think `Metrics` is the only one. It looks like its constructor starts a thread and so on.

But yes, we should shut down all that. Fixed.